scipy/optimize/optimize.py

```diff
@@ -451,17 +451,24 @@ def rosen_hess_prod(x, p):
     return Hp
 
 
-def _wrap_function(function, args):
+def _wrap_scalar_function(function, args):
     # wraps a minimizer function to count number of evaluations
     # and to easily provide an args kwd.
-    # A copy of x is sent to the user function (gh13740)
     ncalls = [0]
     if function is None:
         return ncalls, None
 
     def function_wrapper(x, *wrapper_args):
         ncalls[0] += 1
-        return function(np.copy(x), *(wrapper_args + args))
+        # A copy of x is sent to the user function (gh13740)
+        fx = function(np.copy(x), *(wrapper_args + args))
+        # Ideally, we'd like to a have a true scalar returned from f(x). For
+        # backwards-compatibility, also allow np.array([1.3]), np.array([[1.3]]) etc.
+        try:
+            fx = np.asarray(fx).item()
+        except (TypeError, ValueError) as e:
+            raise ValueError("Objective function must return a scalar") from e
```

Comment:
Let's have the exception here match `ScalarFunction`
```suggestion
            raise ValueError("The user-provided objective function "
                             "must return a scalar value.") from e
```